File: Problem_2.js
// Problem2 Design Autocomplete System (https://leetcode.com/problems/design-search-autocomplete-system/)

// input
// Time Complexity : 
// Avg O(1) 
// When a new entry, O(n), n is the length of sentence
// Initial setup, O(mn), m is total sentences, n is avg length of each sentence
// Space Complexity : O(mn)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class TrieNode {
    constructor() {
        // Maps character and trienode
        this.children = new Map();
        // An array which contains the search output for sentence till char 
        this.pq = [];
    }
    insert(root, freqMap, sentence) {
        let curr = root;
        // Loop through each character of sentence
        for (let i = 0; i < sentence.length; i++) {
            let ch = sentence[i];
            // If it is not present in trie, add trienode for the ch
            if (!curr.children.has(ch)) {
                curr.children.set(ch, new TrieNode());
            }
            // Set curr as the trienode
            curr = curr.children.get(ch);
            // If the result does not contain sentence, add
            if (!curr.pq.includes(sentence))
                curr.pq.push(sentence);
            // Sort by occurence frequency. If same, sort lexicofraphically
            curr.pq.sort((a, b) => {
                if (freqMap.get(b) === freqMap.get(a)) {
                    return a.localeCompare(b);
                }
                return freqMap.get(b) - freqMap.get(a);
            })
            // If size inc by 3, remove from result
            if (curr.pq.length > 3)
                curr.pq.pop();
        }
    }
    search(root, sentence) {
        let curr = root;
        // Loop through each character of sentence and check if it is present in trie
        for (let i = 0; i < sentence.length; i++) {
            let ch = sentence[i];
            // If any ch is not present, return empty
            if (!curr.children.has(ch)) {
                return [];
            }
            curr = curr.children.get(ch);
        }
        // If all ch are present, the curr.pq contains result
        return curr.pq;
    }
}

/**
* @param {string[]} sentences
* @param {number[]} times
*/
class AutocompleteSystem {
    constructor(sentences, times) {
        this.root = new TrieNode();
        this.freqMap = new Map();
        this.stringTillNow = "";
        for (let i = 0; i < sentences.length; i++) {
            // FreqMap maintains sentence and its requency
            this.freqMap.set(sentences[i], times[i]);
            // Insert all sentences in trie
            this.root.insert(this.root, this.freqMap, sentences[i])
        }
    }
};

/** 
* @param {character} c
* @return {string[]}
*/
AutocompleteSystem.prototype.input = function (c) {
    if (c === "#") {
        // Add to freqMap if not present else update
        let str = this.stringTillNow;
        if (!this.freqMap.has(str)) {
            this.freqMap.set(str, 0);
        }
        // Update sentence frequency
        this.freqMap.set(str, this.freqMap.get(str) + 1)
        // Reset string till now.
        this.stringTillNow = "";
        // Update in trie
        this.root.insert(this.root, this.freqMap, str);
        return [];
    }
    // Update string formed
    this.stringTillNow += c;
    // Returns the 1st 3 matches
    return this.root.search(this.root, this.stringTillNow);
};

/** 
* Your AutocompleteSystem object will be instantiated and called as such:
* var obj = new AutocompleteSystem(sentences, times)
* var param_1 = obj.input(c)
*/

File: problem2.cpp

// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
// Time Complexity :O(p*nlog3) n = sentences.size()
// Space Complexity : O(n*p) p=sentence[i].size()
// Did this code successfully run on Leetcode : yes
approach : 1  

class AutocompleteSystem {
public:
    unordered_map<string,int>strmap;
    string s;
    
    AutocompleteSystem(vector<string>& sentences, vector<int>& times) {
        for(int i = 0;i<sentences.size();i++)
            strmap.insert({sentences[i],times[i]});

    }
    
    vector<string> input(char c) {
        
        if(c=='#'){
            strmap[s]++;
            s.clear();
            return {};
        }
        
        s.push_back(c);
        
        
        vector<string> ans;
        
        
        auto cmp = [&](const string& a, const string& b) {
            if (this->strmap[a] == this->strmap[b]) {
                return a < b;
            }
            return this->strmap[a] > this->strmap[b];
        };

        priority_queue<string, vector<string>, decltype(cmp)> pq(cmp);
        
        
        // searching for a prefix (s) in map and put in pq;
        
        for(auto &p: strmap)
        {
            if(p.first.find(s)==0){
                pq.push(p.first);
                if(pq.size()>3){
                    pq.pop();
                }
            }
        }
        
        vector<string>result;

        
        while(!pq.empty())
        {
            result.push_back(pq.top());
            pq.pop();
        }
        reverse(result.begin(),result.end());
        
        return result;
    }
};

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem* obj = new AutocompleteSystem(sentences, times);
 * vector<string> param_1 = obj->input(c);
 */


//Approach 2:  using Trie
// Time Complexity :O(p*n) n = sentences.size() but searching time greatly reduced
// Space Complexity : O(n*p) p = sentence[i].size()
// Did this code successfully run on Leetcode : yes

class AutocompleteSystem {
public:
    unordered_map<string,int>strmap;
    string s;
    class TrieNode{
        public:
        unordered_map<char,TrieNode*>children;
        vector<string>pq;
        TrieNode(){
            
        }
        
    };
    
    TrieNode* root;
    
    void insert(string word,int t)
    {
        TrieNode* curr = root;
        
        for(int i = 0;i<word.size();i++)
        {
            char c = word[i];
            if(curr->children.count(c)==0){
                curr->children.insert({c,new TrieNode()});
            }
            curr = curr->children[c];

            vector<string>vec = curr->pq;
            bool flag = false;
            for(int j = 0;j<vec.size();j++)
            {
                if(vec[j]==word){
                    flag = true;
                    break;
                } 
            }
            if(!flag){
                vec.push_back(word);
            } 
            
            sort(vec.begin(),vec.end(),[&](auto const &a, auto const &b){
                if(strmap[a]==strmap[b]){
                    return a<b;
                }
                return strmap[a]>strmap[b];
            });
            
            if(vec.size()>3){
                vec.pop_back();
            }
            curr->pq = vec;

        }
    }
    
    AutocompleteSystem(vector<string>& sentences, vector<int>& times) {
        root = new TrieNode();
        for(int i = 0;i<sentences.size();i++)
        {
            strmap[sentences[i]]+= times[i];
            insert(sentences[i],times[i]);
        }
    }
    
    vector<string> input(char c) {
        if(c=='#')
        {
            strmap[s]++;
            insert(s,1);
            s.clear();
            return {};
        }
        
        s.push_back(c);
        
        TrieNode* curr = root;
        for(int i = 0;i<s.size();i++)
        {
            if(curr->children.count(s[i])==0){
                return {};
            }
            curr = curr->children[s[i]];
        }
        
        return curr->pq;
        
    }
};

/**
 * Your AutocompleteSystem object will be instantiated and called as such:
 * AutocompleteSystem* obj = new AutocompleteSystem(sentences, times);
 * vector<string> param_1 = obj->input(c);
 */